lib/src/validator/leak_detection.dart

```diff
@@ -105,14 +106,33 @@ final class LeakDetectionValidator extends Validator {
 
 /// Instance of a match against a [LeakPattern].
 final class LeakMatch {
+  final Map<String, SourceFile> sourceFileCache;
+
   final LeakPattern pattern;
-  final SourceSpan span;
 
-  LeakMatch(this.pattern, this.span);
+  final String content;
+  final String url;
+  final int start;
+  final int end;
+
+  SourceSpan span() {
+    final sourceFile =
+        sourceFileCache[url] ??= SourceFile.fromString(content, url: url);
+    return sourceFile.span(start, end);
+  }
```

Comment:
Passing around `sourceFileCache` like this is an ugly hack.

Just remove the `span()` function. This isn't used many places.. Then have the place the prints `LeakMatch` instances do your hack. If that's what you want.

---------------

It might also be reasonable not to compute `SourceFile` when the file is sufficient large.

It might be reasonable to detect binary files and not do `SourceSpan` for such files.

------

Or maybe a factory constructor on `SourceFile` that returns a `_LazySourceFile` which lazily computes the line numbers.

```dart
class _LazySourceFile implements SourceFile {
  final Iterable<int> _decodedChars;
  final Object? _url;
  late final _sourceFile = SourceFile.decoded(_decodedChars, _url);

  _LazySourecFile(this._decodedChars, this._url) {
    // TODO: Some cheap validations of _url and _decodedChars, just handle cases where
    //       SourceFile.decoded would throw.
  }

  @override
  int get lines => _sourceFile.lines;

  // TODO: All other overrides...
}
```

Reply:
I removed the cache, and now only create the sourcefile when "describe"ing the leak.

That way we might create sourcefile multiple times for the same file, but as we at most display 3 leaks it should not be a problem.

I also implemented a limit on when we are creating the sourcefile. Above that we don't attempt to display line numbers, but just offsets.